contracts/modules/Experimental/Checkpoints/PLCRVotingCheckpoint.sol

```diff
@@ -0,0 +1,313 @@
+        uint256 _proposedQuorum,
+        uint256 _totalSupply,
+        uint256 _checkpointId
+    );
+    event VoteRevealed(address _voter, uint256 _weight, uint256 indexed _ballotId, uint256 _choiceOfProposal, uint256 _salt, bytes32 _secretVote);
+    event BallotStatusChanged(uint256 _ballotId, bool _newStatus);
+
+    constructor(address _securityToken, address _polyAddress)
+    public
+    Module(_securityToken, _polyAddress)
+    {
+
+    } 
+
+    /**
+     * @notice Use to create the ballot
+     * @param _commitDuration Unix time period till the voters commit there vote
+     * @param _revealDuration Unix time period till the voters reveal there vote starts when commit duration ends
+     * @param _noOfProposals Total number of proposal used in the ballot. In general it is 2 (For & Against)
+     * @param _proposedQuorum Minimum number of weight vote requires to win a election.
+     */
+    function createBallot(uint256 _commitDuration, uint256 _revealDuration, uint256 _noOfProposals, uint256 _proposedQuorum) external withPerm(ADMIN) {
+        uint256 startTime = now;
+        uint256 checkpointId = ISecurityToken(securityToken).createCheckpoint();
+        _createBallotWithCheckpoint(_commitDuration, _revealDuration, _noOfProposals, _proposedQuorum, checkpointId, startTime);
+    }
+
+    /**
+     * @notice Use to create the ballot
+     * @param _commitDuration Unix time period till the voters commit there vote
+     * @param _revealDuration Unix time period till the voters reveal there vote starts when commit duration ends
+     * @param _noOfProposals Total number of proposal used in the ballot. In general it is 2 (For & Against)
+     * @param _proposedQuorum Minimum number of weight vote requires to win a election.
+     * @param _checkpointId Valid checkpoint Id
+     * @param _startTime startTime of the ballot
+     */
+    function createCustomBallot(
+        uint256 _commitDuration,
+        uint256 _revealDuration,
+        uint256 _noOfProposals,
+        uint256 _proposedQuorum,
+        uint256 _checkpointId,
+        uint256 _startTime
+    )
+        external
+        withPerm(ADMIN)
+    {
+        require(_checkpointId <= ISecurityToken(securityToken).currentCheckpointId(), "Invalid checkpoint Id");
+        _createBallotWithCheckpoint(_commitDuration, _revealDuration, _noOfProposals, _proposedQuorum, _checkpointId, _startTime);
+    }
+
+    function _createBallotWithCheckpoint(
+        uint256 _commitDuration,
+        uint256 _revealDuration,
+        uint256 _totalProposals,
+        uint256 _proposedQuorum,
+        uint256 _checkpointId,
+        uint256 _startTime
+    )
+        internal
+    {
+        // Sanity checks
+        _validValueCheck(_commitDuration);
+        _validValueCheck(_revealDuration);
+        _validValueCheck(_proposedQuorum);
+        require(_startTime >= now, "Invalid start time");
+        require(_totalProposals > 1, "totalProposals should be > 1");
```

Comment:
For consistency, reason strings should mention what went wrong rather than what is the valid input.